### ISSUE TITLE
Machine Learning - Optimizer

### ISSUE THREAD

**Issue by mbianchif** [lminervino18/oxidized-neural-orchestra#133]
Only let the Parameter Server and AllReduce trainer select a different optimizer other than gradient descent.

A `Trainer` should only be able to use gradient descent.

When training under Parameter Server the other stateful optimizer's state is not being shared between the nodes, just make the trainers stateless optimizers by using gradient descent and letting the server use stateful optimizers.

When training under All Reduce whenever the `Trainer` is used to compute the next step gradient it would use the inner optimizer buffers for the stateful optimizers, meaning that whenever the total gradient optimization is meant to be done, these buffers now contain garbage.
